### PR TITLE
bpo-32348: Optimize asyncio.Future schedule/add/remove callback.

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue32348

### CODE REVIEW

**Comment by asvetlov** [comment]
After the change C future has `callback0` member but Python version has not.
Maybe adding `callback0` to py future for sake of diversity reducing makes sense?

**Comment by 1st1** [comment]
I actually want to keep Python implementation simple (for educational purposes) and different (to catch any difference in semantics).

**Comment by asvetlov** [comment]
Got it. Ok.

File: Modules/_asynciomodule.c [hunk 1]
```diff
@@ -416,25 +478,64 @@ future_get_result(FutureObj *fut, PyObject **result)
...
-        int err = PyList_Append(fut->fut_callbacks, arg);
-        if (err != 0) {
-            return NULL;
+        else if (fut->fut_callback0 == NULL) {
+            Py_INCREF(arg);
+            fut->fut_callback0 = arg;
+        }
+        else {
+            assert(fut->fut_callback0 != NULL);
+            assert(fut->fut_callbacks == NULL);
```
Comment (methane) [inline Modules/_asynciomodule.c]: These asserts seems too redundant.

**Review by asvetlov — APPROVED** [review]

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -1122,12 +1328,33 @@ typedef struct {
     FutureObj *future;
 } futureiterobject;
 
+
+#define FI_FREELIST_MAXLEN 100
+static futureiterobject *fi_freelist_head = NULL;
+static futureiterobject *fi_freelist_tail = NULL;
+static Py_ssize_t fi_freelist_len = 0;
```
Comment (methane) [inline Modules/_asynciomodule.c]: I recommend LIFO than FIFO when implementing freelist.
It may be better CPU cache ratio.

**Review by methane — APPROVED** [review]

File: Modules/_asynciomodule.c (hunk 1 above)
Comment (1st1) [reply Modules/_asynciomodule.c]: dropped them

File: Modules/_asynciomodule.c (hunk 2 above)
Comment (1st1) [reply Modules/_asynciomodule.c]: changed to LIFO

**Pull request merged**